<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import plotly.express as px

In [67]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select count(id) as vacancies
from public.vacancies
'''
# общее количество вакансий

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


vacancies
0      49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select count(id) as employers
from public.employers
'''
# общее количество работодателей

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


employers
0      23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(id) as areas
from public.areas
'''
# общее количество регионов

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


areas
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(id) as industries
from public.industries
'''
# общее количество сфер деятельности

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


industries
0         294

***

In [12]:
# выводы по предварительному анализу данных

В процессе выполнения заданий блока стало известно, что общее количество вакансий составляет 49197, число работодателей 23501, количество регионов размещения вакансии 1362 и число сфер деятельности 294.

Исходя из результатов, полученных с помощью запросов, можно сделать выводы о том, что база данных предлагает вакансии для широкого спектра профессий и специальностей, а также имеет крупный географический охват. В среднем на один регион приходится 36 вакансий в 17 кампаниях, хотя конечно же реальное распределение будет отличаться неоднородностью фактических данных.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''
select a.name as area, count(v.id) as vacancies
from public.areas a
join public.vacancies v on a.id = v.area_id 
group by 1
order by 2 desc
limit 5
'''
# присоединение к таблице регионов таблицы с вакансиями с последующей группировкой по регионам в порядке убывания количества вакансий

In [14]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area  vacancies
0           Москва       5333
1  Санкт-Петербург       2851
2            Минск       2112
3      Новосибирск       2006
4           Алматы       1892

In [15]:
barplot = px.bar(data_frame=df, x='area', y='vacancies', text='vacancies', title='Диаграмма 4.1 - Топ-5 регионов по количеству вакансий', \
    height=500, width=1000, labels={'vacancies':'Количество вакансий', 'area':'Регион'})
barplot
# столбчатая диаграмма: топ-5 регионов по количеству вакансий

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
# текст запроса
query_4_2 = f'''
select count(id) as vacancies
from public.vacancies
where salary_from is not null or salary_to is not null
'''
# фильтрация по наличию заполненных полей зарплат

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\970655580.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



vacancies
0      24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
# текст запроса
query_4_3 = f'''
select round(avg(salary_from)) as from, round(avg(salary_to)) as to
from public.vacancies
'''
# средние значения для полей зарплаты, округленные до целого

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2500607024.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



from        to
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [20]:
# текст запроса
query_4_4 = f'''
select distinct schedule, employment, count(id) as vacancies
from public.vacancies
group by 1, 2
order by 3 desc
'''
# количество вакансий для каждого сочетания графика и типа трудоустройства в порядке убывания

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2022935265.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



schedule           employment  vacancies
0        Полный день     Полная занятость      35367
1   Удаленная работа     Полная занятость       7802
2      Гибкий график     Полная занятость       1593
3   Удаленная работа  Частичная занятость       1312
4     Сменный график     Полная занятость        940
5        Полный день           Стажировка        569
6     Вахтовый метод     Полная занятость        367
7        Полный день  Частичная занятость        347
8      Гибкий график  Частичная занятость        312
9        Полный день     Проектная работа        141
10  Удаленная работа     Проектная работа        133
11     Гибкий график           Стажировка        116
12    Сменный график  Частичная занятость        101
13  Удаленная работа           Стажировка         64
14     Гибкий график     Проектная работа         18
15    Сменный график           Стажировка         12
16    Вахтовый метод     Проектная работа          2
17    Сменный график     Проектная работа          1

In [22]:
# тепловая карта количества вакансий в зависимости от графика и типа трудоустроства
pivot = df.pivot_table(values='vacancies', columns='employment', index='schedule')
heatmap = px.imshow(pivot, labels={'schedule':'График работы', 'employment':'Тип трудоустройства'},\
    title='Диаграмма 4.4 - Тепловая карта вакансий', height=400, width=800)
heatmap

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [23]:
# текст запроса
query_4_5 = f'''
select experience, count(id) as vacancies
from public.vacancies
group by 1
order by 2
'''
# количество вакансий, сгрупированные по категориям опыта работы

In [24]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2316260449.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  vacancies
0         Более 6 лет       1337
1           Нет опыта       7197
2       От 3 до 6 лет      14511
3  От 1 года до 3 лет      26152

In [25]:
#круговая диаграмма: число вакансий для категорий опыта работы
pie = px.pie(data_frame=df, names='experience', values='vacancies', title='Диаграмма 4.5 - Требуемый опыт работы', height=500, width=800, \
    labels={'experience':'Опыт работы', 'vacancies':'Количество вакансий'})
pie

***

In [26]:
# выводы по детальному анализу вакансий

В прошлом юните мы выяснили, что база вакансий имеет широкую географию. Оценить это нам поможет выборка количества вакансий по каждому региону. На диаграмме 4.1 мы видим, что лидируют Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Однако в конце списка мы увидим неизвестные отдаленные города с количеством вакансий 1 на регион. Конечно, объемы работы, компаний и соискателей распределены неоднородно и концентрируются вблизи экономических центров и крупных городов.

Отдельно хочется упомянуть вакансии с незаполненными полями зарплат. При общем количестве вакансий почти 50 тысяч, хотя бы одно поле (зарплата от, зарплата до) заполнено лишь у половины. Возможно это удобно работодателю, но усложняет процесс поиска и анализа позиций с таким количеством незаполненных полей зарплат.

В зависимости от таких параметров вакансии, как график работы и тип трудоустройства, распределение количества вакансий наглядно видно на тепловой карте (Диаграмма 4.4): так с большим перевесом преобладает сочетание полной занятости и полного рабочего дня, на втором месте - полная занятость и удаленная работа. А далее различия между категориями не так заметны. Также бросается в глаза отсутсвие некоторых данных. Вахтовый метод работы не предполагает стажировку или частичную занятость.

Если обратить внимание на требуемый опыт работы (Диаграмма 4.5), видно, поиск начинающих сотрудников (от 1 до 3 лет) превышает поиск опытных специалистов (более 6 лет) почти в 20 раз. Этот наиболее крупный сектор вакансий от общего числа, вероятно, привлекателен для работодатей по соотншению опыта работы и ожидаемой зарплаты соискателя.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [27]:
# текст запроса
query_5_1 = f'''
(select e.name as employer, count(v.id) as vacancies
from public.employers e
join public.vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
limit 1)

union all

(select e.name, count(v.id)
from public.employers e
join public.vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
offset 4
limit 1)
'''
# 1 и 5 место среди работодателей по количеству вакансий
# Присоединение к таблице работодателей таблицы с вакансиями, группировка по работодателю, сортировка по убыванию количества вакансий
# Соединение полученных строк для 1 и 5 мест

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\3834399874.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer  vacancies
0         Яндекс       1933
1  Газпром нефть        331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [29]:
# текст запроса
query_5_2 = f'''
select distinct a.name as area, count(e.id) as employers, count(v.id) as vacancies
from public.areas a
full outer join public.employers e on a.id = e.area
full outer join public.vacancies v on a.id = v.area_id
group by 1
order by 3 nulls first, 2 desc
limit 1
'''
# количество работодателей и вакансий по регионам, результат: регион без вакансий с наибольшим количеством работодателей
# Присоединение к таблице регионов таблиц с работодателями и вакансиями, группировка по регионам, сортировка по вакансиям (nulls first), 
# а затем по работодателям в порядке убывания

In [30]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2250345976.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



area  employers  vacancies
0  Россия        410          0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [31]:
# текст запроса
query_5_3 = f'''
select distinct e.name as employer, count(distinct v.area_id) as areas
from public.employers e
join public.vacancies v on e.id = v.employer_id
group by 1
having count(v.area_id) is not null
order by 2 desc
'''
# количество регионов размещения вакансий по работодателям в порядке убывания
# присоединение к таблице работодателей таблицы с вакансиями, группировка по названию кампании

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\201667002.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer  areas
0                       Яндекс    181
1                   Ростелеком    152
2                   Спецремонт    116
3       Поляков Денис Иванович     88
4                    ООО ЕФИН      71
...                        ...    ...
14761     Яроцкий Е. В./ MoonX      1
14762  Ясли-сад №374 г. Минска      1
14763                      ЯСП      1
14764        Ястро-Переработка      1
14765                     ЯТСС      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [33]:
# текст запроса
query_5_4 = f'''
select count(e.id) as employers
from public.employers e
left join public.employers_industries ei on e.id = ei.employer_id
where ei.employer_id is null
'''
# количество работодателей с пустым полем сферы деятельности
# присоединение к таблице работодателей таблицы работодатель/сферы деятельности с фильтрацией по пустым значениям

In [34]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\3493105328.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employers
0       8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [35]:
# текст запроса
query_5_5 = f'''
select e.name as employer
from public.employers e
join public.employers_industries i on e.id = i.employer_id
group by 1
having count(i.industry_id)=4
order by 1
offset 2
limit 1
'''
# название кампании с 4 сферами деятельности на третьем месте по алфавиту
# присоединение к таблице работодателей таблицы работодатель/сферы деятельности с группировкой по работодателю и фильтрацией по количеству сфер

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\388711819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer
0     2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [37]:
# текст запроса
query_5_6 = f'''
select count(e.id) as employers
from public.employers e
join public.employers_industries e_i on e.id = e_i.employer_id
join public.industries i on e_i.industry_id = i.id
where i.name in ('Разработка программного обеспечения')
'''
# количество кампаний со сферой деятельности "Разработка программного обеспечения"
# присоединение к таблице работодателей таблиц работодатель/сферы деятельности и сферы деятельности с фильтрацией по значению названия сферы

In [38]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\821628714.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employers
0       3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [39]:
# код для получения списка городов-милионников
from bs4 import BeautifulSoup 
import requests                    # импорт библиотек

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8#%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA_%D0%B3%D0%BE%D1%80%D0%BE%D0%B4%D0%BE%D0%B2-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D0%BE%D0%B2'
response = requests.get(url)                                 # отправка веб-запроса
page = BeautifulSoup(response.text, 'html.parser')                 # создание объекта BeautifulSoup из страницы
table = page.find('table', class_='standard sortable')                    # получение данных таблицы

tag_list = []
for i in table.find_all('a'):
    city = i.text                                                  # создание списка значений по тегу
    tag_list.append(city)    

cities = tag_list[4:18]+tag_list[-2:]                           # выбор городов-миллионников
cities


['Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград']

In [40]:
# текст запроса
query_5_7 = f'''
select a.name as area, count(v.id) as vacancies
from public.vacancies v
join public.areas a on a.id = v.area_id
join public.employers e on e.id = v.employer_id
where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск',
'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
group by 1

union all

select 'total', count(v.id)
from public.vacancies v
join public.areas a on a.id = v.area_id
join public.employers e on e.id = v.employer_id
where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск',
'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
order by 2 desc
'''
# регионы и количество вакансий в них компании Яндекс
# присоединение к таблице вакансий таблиц регионов и работодателей, фильтрация по названиям компании и городов с группировкой по региону
# присоединение строки с суммарным количеством вакансий компании, сортировка всей таблицы в порядке убывания вакансий

In [41]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\1965596796.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



area  vacancies
0             total        485
1            Москва         54
2   Санкт-Петербург         42
3      Екатеринбург         39
4   Нижний Новгород         36
5       Новосибирск         35
6           Воронеж         32
7         Краснодар         30
8            Самара         26
9               Уфа         26
10            Пермь         25
11           Казань         25
12   Ростов-на-Дону         25
13        Волгоград         24
14       Красноярск         23
15        Челябинск         22
16             Омск         21

In [42]:
# столбчатая диаграмма: количество вакансий компании Яндекс в городах-миллионниках
query_bar = f'''
select a.name as area, count(v.id) as vacancies
from public.vacancies v
join public.areas a on a.id = v.area_id
join public.employers e on e.id = v.employer_id
where e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск',
'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')
group by 1
order by 2 desc'''
df = pd.read_sql_query(query_bar, connection)
df


barplot = px.bar(data_frame=df, x='area', y='vacancies', labels={'vacancies':'Количество вакансий', 'area':'Город'}, \
    title='Диаграмма 5.7 - Вакансии Яндекс в городах-миллионниках', height=500, width=1000)
barplot

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2443652583.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



***

In [43]:
# выводы по анализу работодателей

Работодатели в базе вакансий тоже представлены разные. Например, по количеству вакансий есть лидер - Яндекс - у которого 1933 вакансии с одной стороны и мелкие компании с 1 вакансией с другой.

Также и по регионам распределение неравномерное. Само деление на регионы в базе вызывает вопросы, потому как и страны, и регионы и города находятся в одной категории, что может привести к ложным выводам. На мой взгляд необходимы отдельные признаки для стран, областей и городов для понятной локализации. В задании 5.2 мы видим, что регион с наибольшим количеством работодателей и нулевым значением вакансий - Россия. Но это не значит, что во всей России 410 работодателей и 0 вакансий на самом деле.
Однако, когда под регионом подразумевается город - все более стабильно. В задании 5.7 подсчитано количество вакансий компании Яндекс в городах миллионниках по отдельности и общее число. Здесь мы видим адекватные значения, лидеры - Москва и Санкт-Петербург и так далее.

По количеству регионов, в которых работодатели размещают вакансии, можно судить об охвате компании. По этому параметру опять таки лидирует Яндекс - 181 регион, затем идет Ростелеком, Спецремонт и далее по ниспадающей. В конце списка локальные компании, работающие только в своем регионе. 

Касательно сфер дефтельности организации можно сказать следующее: конечно работодатель не обязан указывать в вакансии сферу. Таких было насчитано 8419 компании. Однако для фильтрации и подбора нужных вакансий это важный фактор, как, например, в задании 5.6. С точки зрения соискателя, если видов деятельности у организации несколько, то это может означать масштабность и стабильность компании в целом, большое количество различных рабочих мест и комфортные условия труда.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [44]:
# текст запроса
query_6_1 = f'''
select count(v.id) as vacancies
from public.vacancies v
where (lower(v.name) like '%data%') or (lower(v.name) like '%данн%') 
'''
# количество вакансий, название которых содержит 'data', 'данн'. Применена фильтрация по имени вакансий

In [45]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\366702258.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



vacancies
0       1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [46]:
# текст запроса
query_6_2 = f'''
select count(id) as vacancies
from public.vacancies
where 
((name ilike '%data scientist%') or 
(name ilike '%data science%') or 
(name ilike '%исследователь данных%') or 
(name like '%ML%' and name not ilike '%HTML%') or 
(name ilike '%machine learning%') or 
(name ilike '%машинн%обучен%')) and 
((name ilike '%junior%') or (experience = 'Нет опыта') or (employment = 'Стажировка'))
'''
# количество вакансий для junior DS
# фильтрация по названиям вакансий, опыту работы и типу трудоустройства

In [47]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2267038862.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



vacancies
0         51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [48]:
# текст запроса
query_6_3 = f'''
select count(id) as vacancies
from public.vacancies
where 
((name ilike '%data scientist%') or 
(name ilike '%data science%') or 
(name ilike '%исследователь данных%') or 
(name like '%ML%' and name not ilike '%HTML%') or 
(name ilike '%machine learning%') or 
(name ilike '%машинн%обучен%')) and
(key_skills ilike '%SQL%' or key_skills ilike '%postgres%')
'''
# количество вакансий для DS с ключевыми навыками SQL или postgres
# фильтрация по названиям вакансий и ключевых навыков

In [49]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\3794299702.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



vacancies
0        201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [50]:
# текст запроса
query_6_4 = f'''
select count(id) as vacancies
from public.vacancies
where 
((name ilike '%data scientist%') or 
(name ilike '%data science%') or 
(name ilike '%исследователь данных%') or 
(name like '%ML%' and name not ilike '%HTML%') or 
(name ilike '%machine learning%') or 
(name ilike '%машинн%обучен%')) and
(key_skills ilike '%Python%')
'''
# количество вакансий для DS с ключевым навыком Python
# фильтрация по названиям вакансий и ключевых навыков

In [51]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\4264665179.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



vacancies
0        351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [52]:
# текст запроса
query_6_5 = f'''
select round(avg((length(key_skills)-length(replace(key_skills, '\t', ''))+1)), 2) as mean
from public.vacancies
where 
((name ilike '%data scientist%') or 
(name ilike '%data science%') or 
(name ilike '%исследователь данных%') or 
(name like '%ML%' and name not ilike '%HTML%') or 
(name ilike '%machine learning%') or 
(name ilike '%машинн%обучен%'))
'''
# среднее количество ключевых навыков в вакансии для DS
# применение операторов для работы со строковыми и числовыми данными, вычисление числа навыков исходя из количества разделителей, 
# подсчет среднего значения и округление результата, фильтрация по названиям вакансий

In [53]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\3654210517.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



mean
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [54]:
# текст запроса
query_6_6 = f'''
select distinct experience, round((avg(coalesce(salary_from, salary_to)) + avg(coalesce(salary_to, salary_from)))/2) as salary
from public.vacancies
where 
((salary_from is not null or salary_to is not null) and 
(name ilike '%data scientist%') or 
(name ilike '%data science%') or 
(name ilike '%исследователь данных%') or 
(name like '%ML%' and name not ilike '%HTML%') or 
(name ilike '%machine learning%') or 
(name ilike '%машинн%обучен%')) and
experience = 'От 3 до 6 лет'
group by 1
'''
# средняя зарплата для DS с опытом работы от 3 до 6 лет
# вычисление средней заплаты, исходя из того, сколько полей заполнено и округление результат до целого
# фильтрация по названию вакансий, опыту работы и заполненности полей зарплаты, группировка по опыту работы

In [55]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\519194770.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience    salary
0  От 3 до 6 лет  243115.0

***

In [56]:
# выводы по предметному анализу

В данном разделе были детально изучены вакансии, связанные с данными и DS. Их общее количество (1771 вакансия), вакансии для начинающих DS всего 51), вакансии с конкретными ключевыми навыками, а также средний размер оплаты труда для специалистов с различным опытом работы.
В целом, данное исследование демонстрирует алгоритм подбора вакансий под определенные условия поиска. Были применены фильтрация по названиям вакансий, ключевым навыкам, графику работы, типу трудоустройства, опыту работы и средней зарплате.

Если в качестве примера брать специалиста DS с опытом работы от 3 до 6 лет, то средняя оплата труда более,чем достойная. Но при этом соискатель должен обладать обширными знаниями и большим количеством ключевых навыков. То есть требования работодателя соответсвенные.

# Общий вывод по проекту

In [57]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Подводя итог, хочется сказать, что данные, по которым производилось исследование, позволяют работать с базой вакансий с целью подбора подходящих вакансий на основании заданных условий путем создания соответсвующих запросов. Задания были выполнены, результаты поиска и подсчета получены.
В плане доработки этой базы данных считаю нужным разделить регионы на соответсвующие категории объектов.

В качестве дополнительных исследований хотелось бы привести запросы, интересные лично мне:
1. Как меняется средняя зарплата DS без опыта и с оптытом от 1 до 3 лет в зависимости от города
2. Какие ключевые навыки требуются на первых этапах трудоустройства в крупных компаниях

In [58]:
dop = f'''
select v.experience, a.name as city, 
round((avg(coalesce(v.salary_from, v.salary_to)) + avg(coalesce(v.salary_to, v.salary_from)))/2) as salary
from public.vacancies v
join public.areas a on a.id = v.area_id
where
((v.name ilike '%data scientist%') or 
(v.name ilike '%data science%') or 
(v.name ilike '%исследователь данных%') or 
(v.name like '%ML%' and v.name not ilike '%HTML%') or 
(v.name ilike '%machine learning%') or 
(v.name ilike '%машинн%обучен%')) and
((v.experience = 'От 1 года до 3 лет') or (v.experience = 'Нет опыта')) and
((a.name = 'Москва') or (a.name = 'Краснодар'))
group by 1, 2
order by 3
'''

In [59]:
df = pd.read_sql_query(dop, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\1051882271.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience       city    salary
0           Нет опыта     Москва   72500.0
1  От 1 года до 3 лет     Москва  169179.0
2  От 1 года до 3 лет  Краснодар       NaN

Прогнозы неутешительные. Согласно этой базе данных, в Краснодаре без опыта сотрудники не требуются, а для соискателей со стажем от 1 до 3 лет зарплата неизвестна. Зато в Москве перспективы есть. Добавим к нашему запросу Ростов, чтобы не так далеко переезжать и добавим количество самих вакансий, чтобы примерно оценить конкурс на место.

In [60]:
dop = f'''
select 
v.experience, 
a.name as city, 
round((avg(coalesce(v.salary_from, v.salary_to)) + avg(coalesce(v.salary_to, v.salary_from)))/2) as salary, 
count(v.id) as vacancies
from public.vacancies v
join public.areas a on a.id = v.area_id
where
((v.name ilike '%data scientist%') or 
(v.name ilike '%data science%') or 
(v.name ilike '%исследователь данных%') or 
(v.name like '%ML%' and v.name not ilike '%HTML%') or 
(v.name ilike '%machine learning%') or 
(v.name ilike '%машинн%обучен%')) and
((v.experience = 'От 1 года до 3 лет') or (v.experience = 'Нет опыта')) and
(a.name = 'Москва' or a.name = 'Краснодар' or a.name = 'Ростов-на-Дону')
group by 1, 2
order by 1, 3
'''

In [61]:
df = pd.read_sql_query(dop, connection)
df

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\1051882271.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience            city    salary  vacancies
0           Нет опыта          Москва   72500.0         10
1  От 1 года до 3 лет  Ростов-на-Дону   50000.0          5
2  От 1 года до 3 лет          Москва  169179.0        107
3  От 1 года до 3 лет       Краснодар       NaN          4

Так видно, что Ростов не так уж и привлекателен для переезда. И в целом, количество вакансий для соискателей с минимальным опытом в Москве и Краснодаре/Ростове различается в десятки раз. Ну что ж, посмотрим какие ключевые навыки мне понадобятся.

In [62]:
dop = f'''
select e.name as employer, replace(key_skills, '\t', ', ') as "key skills"
from public.employers e
join public.vacancies v on e.id = v.employer_id
where
((v.name ilike '%data scientist%') or 
(v.name ilike '%data science%') or 
(v.name ilike '%исследователь данных%') or 
(v.name like '%ML%' and v.name not ilike '%HTML%') or 
(v.name ilike '%machine learning%') or 
(v.name ilike '%машинн%обучен%')) and
(v.experience = 'Нет опыта') 
group by 1,2
order by count(v.id) desc
'''

In [63]:
df1 = pd.read_sql_query(dop, connection)
df1

C:\Users\Anna\AppData\Local\Temp\ipykernel_11060\2655579351.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer  \
0                                              Comexp   
1                                       SteadyControl   
2                                   DNS Головной офис   
3                                                Ozon   
4    The Skolkovo Institute of Science and Technology   
5    The Skolkovo Institute of Science and Technology   
6                             Алгоритмические Системы   
7                                          АЛЬФА-БАНК   
8                                             Апэрбот   
9                           БСТ Менеджмент-Консалтинг   
10                                             Диплей   
11                                             Диплей   
12  КБ Уральский банк реконструкции и развития (УБ...   
13                                            МегаФон   
14                           Миррико, Группа Компаний   
15                                           ННФормат   
16                                               СБЕР   
17      СтарЛайн, Научно-производственное объединение   
18                                  Строительный Двор   
19                                      Улыбка радуги   
20                                   Университет ИТМО   
21  Федеральное автономное учреждение Государствен...   
22                                   Целус Менеджмент   
23                                          Шлюмберже   
24                                               DINS   

                                           key skills  
0   Machine Learning, Deep Learning, Data Science,...  
1   Python, Unix, Аналитическое мышление, Нейронны...  
2   Python, Linux, Git, SQL, Docker, Bash, OpenCV,...  
3   Python, SQL, PostgreSQL, ML, Machine Learning,...  
4   Python, PyTorch, Git, Tensorflow, Machine Lear...  
5   PyTorch, Machine Learning, Конференции, Python...  
6   Английский язык, CJM, Подготовка презентаций, ...  
7                                                None  
8          Python, Django Framework, SQL, XML, MS SQL  
9                                          Python, ML  
10  Python, SQL, Java, Математический анализ, Мате...  
11  Python, SQL, Java, Математический анализ, Мате...  
12  Python, SQL, Data Mining, C++, Математический ...  
13                                 SQL, Oracle Pl/SQL  
14  ML, линейные модели, Random Forest, Gradient B...  
15  Python, Математическая статистика, Прогнозиров...  
16                                               None  
17                                      Linux, Python  
18  Python, Machine Learning, Pandas, SQL, Numpy, ...  
19  Python, SQL, Machine Learning, Data Science, А...  
20  Python, Machine Learning, Математическое модел...  
21               Python, ML, Pytorch, Computer Vision  
22  Python, C++, Оптимизация кода, Анализ кода, Ма...  
23  Python, Git, TensorFlow, Keras, Jupiter, NLP, ...  
24  Computer Vision, PyTorch, Tensorflow, Deep Lea...

Благодаря такой выборке становится легко определить направления роста и развития как специалиста. Как говорится, путь осилит идущий.
 
В качестве заключения хочется отметить, что с помощью SQL-запросов удобно работать с БД и извлекать из них необходимую информацию. 

In [64]:
connection.close()